compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp

```diff
@@ -0,0 +1,424 @@
+    }
+    auto linalgOp = dyn_cast<linalg::LinalgOp>(op);
+    if (!linalgOp) {
+      return false;
+    }
+    if (!linalg::isaContractionOpInterface(linalgOp)) {
+      return false;
+    }
+    if (groupedOperations.contains(linalgOp) || allOps.contains(linalgOp)) {
+      return false;
+    }
+    if (linalgOp->getOperand(0).getType() != lhsType ||
+        linalgOp->getOperand(1).getType() != rhsType ||
+        linalgOp->getOperand(2).getType() != outType) {
+      return false;
+    }
+    // Either this has to dominate the seed or seed has to dominate this.
+    if (!dominanceInfo.properlyDominates(linalgOp, seedOp) &&
+        !dominanceInfo.properlyDominates(seedOp, linalgOp)) {
+      return false;
+    }
+    return true;
+  };
+
+  // Iterating over the users as is bad cause the users have no ordering
+  // gaurantees. So look for only ops within the same block as the seed op and
+  // sort them.
+  SmallVector<Operation *> lhsUsers;
+  for (Operation *lhsUser : lhs.getUsers()) {
+    if (lhsUser->getBlock() == seedOp->getBlock()) {
+      lhsUsers.push_back(lhsUser);
+    }
+  }
+  llvm::sort(lhsUsers, [&](Operation *lhs, Operation *rhs) {
+    return dominanceInfo.properlyDominates(lhs, rhs);
+  });
+
+  // Collect all contraction op users of lhs.
+  for (Operation *lhsUser : lhsUsers) {
+    if (lhsUser == seedOp.getOperation()) {
+      continue;
+    }
+
+    auto linalgUser = dyn_cast<linalg::LinalgOp>(lhsUser);
+    if (!canBeGrouped(linalgUser)) {
+      continue;
+    }
+
+    std::optional<linalg::GenericOp> userTruncOp = getTruncUser(linalgUser);
+    if (truncateOps && !userTruncOp) {
+      continue;
+    }
+
+    contractionOps.push_back(linalgUser);
+    allOps.insert(linalgUser);
+    if (truncateOps) {
+      truncateOps.value().push_back(userTruncOp.value());
+      allOps.insert(userTruncOp.value());
+    }
+  }
+
+  if (contractionOps.size() == 1) {
+    return std::nullopt;
+  }
+
+  // Check that all uses are "self-contained", i.e. one of the ops dominates
+  // all the ops (for both the contraction and the truncation ops)
+  // and that this op dominates all the other uses of all the ops.
+  std::optional<Operation *> dominatedByAllOps =
+      findDominatedByAllOps(allOps, dominanceInfo);
+  if (!dominatedByAllOps) {
+    return std::nullopt;
+  }
+
+  // Check that this operation dominates all other uses of allOps.
+  for (auto op : allOps) {
+    if (op == dominatedByAllOps.value()) {
+      continue;
+    }
+    for (Operation *user : op->getUsers()) {
+      if (allOps.contains(user)) {
+        continue;
+      }
+      if (!dominanceInfo.properlyDominates(dominatedByAllOps.value(), user)) {
+        return std::nullopt;
+      }
+    }
+  }
+  return HorizontalFusionGroup{contractionOps, truncateOps,
+                               dominatedByAllOps.value()};
+}
+
+static LogicalResult fuseGroup(RewriterBase &rewriter,
+                               HorizontalFusionGroup &fusionGroup) {
```

Comment:
Please add docs and examples for the method because it is not trivial. Also, inlining comments in the implementation can help readability.

I did not expect tensor ops (like concat, reshapes) generated in the pass. I thought that it could create a `flow.dispatch.*` and put all the operations to the dispatch. I realized that it is not the case after I reviewed the code. So I think doc would be very helpful.

Reply:
Added